service/src/main/java/javasabr/mqtt/service/publish/handler/impl/Qos1MqttPublishInMessageHandler.java

```diff
@@ -32,29 +34,68 @@ public QoS qos() {
+    if (alreadyInProcess != null) {
+      // in the case if we already process the fist publish attempt, we can skip it
+      if (publish.duplicated() && alreadyInProcess.messageType() == MqttMessageType.PUBLISH) {
+        return false;
+      }
+      handleMessageIdIsInUse(client, messagedId);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected void handleNoMatchedSubscribers(
+      ExternalMqttClient client,
+      MqttSession session,
+      Publish publish) {
+    super.handleNoMatchedSubscribers(client, session, publish);
+    int messageId = publish.messageId();
+    MqttOutMessage response = messageOutFactoryService
         .resolveFactory(client)
-        .newPublishAck(publish.messageId(), PublishAckReasonCode.NO_MATCHING_SUBSCRIBERS));
+        .newPublishAck(messageId, PublishAckReasonCode.NO_MATCHING_SUBSCRIBERS);
+    sendFeedback(client, session, response, messageId);
   }
 
   @Override
-  protected void handleError(ExternalMqttClient client, Publish publish, PublishHandlingResult handlingResult) {
-    super.handleError(client, publish, handlingResult);
-    log.debug(client.clientId(), "[%s] Send PUBACK after failed processing publish..."::formatted);
-    client.send(messageOutFactoryService
+  protected void handleSuccess(
+      ExternalMqttClient client,
+      MqttSession session,
+      Publish publish,
+      int matchedSubscribers) {
+    super.handleSuccess(client, session, publish, matchedSubscribers);
+    int messageId = publish.messageId();
+    MqttOutMessage response = messageOutFactoryService
         .resolveFactory(client)
-        .newPublishAck(publish.messageId(), handlingResult.ackReasonCode()));
+        .newPublishAck(messageId, PublishAckReasonCode.SUCCESS);
+    sendFeedback(client, session, response, messageId);
   }
 
   @Override
-  protected void handleSuccessfulResult(ExternalMqttClient client, Publish publish, int subscribers) {
-    super.handleSuccessfulResult(client, publish, subscribers);
-    log.debug(client.clientId(), "[%s] Send PUBACK after successful processing publish..."::formatted);
+  protected void handleError(
+      ExternalMqttClient client,
+      MqttSession session,
+      Publish publish,
+      PublishHandlingResult handlingResult) {
+    super.handleError(client, session, publish, handlingResult);
+    int messageId = publish.messageId();
+    MqttOutMessage response = messageOutFactoryService
+        .resolveFactory(client)
+        .newPublishAck(publish.messageId(), handlingResult.ackReasonCode());
+    sendFeedback(client, session, response, messageId);
+  }
+
+  private void handleMessageIdIsInUse(ExternalMqttClient client, int messageId) {
     client.send(messageOutFactoryService
         .resolveFactory(client)
-        .newPublishAck(publish.messageId(), PublishAckReasonCode.SUCCESS));
+        .newPublishAck(messageId, PublishAckReasonCode.PACKET_IDENTIFIER_IN_USE));
```

Comment:
Could we consider renaming PACKET_IDENTIFIER_IN_USE to MESSAGE_IDENTIFIER_IN_USE ?